usim/py/resources/resource.py

```diff
@@ -0,0 +1,219 @@
+        # request is no longer held here
+    """
+    #: undocumented, set by Resource._do_get
+    __slots__ = 'usage_since',
+
+    def __exit__(self, exc_type, value, traceback):
+        if self.triggered:
+            self.resource.release(self)
+        super().__exit__(exc_type, value, traceback)
+
+
+class Release(Get):
+    """
+    Release a previous ``request`` of a ``resource``
+    """
+    __slots__ = 'request',
+
+    def __init__(self, resource, request):
+        self.request = request
+        super(Release, self).__init__(resource)
+
+
+class Resource(BaseResource):
+    r"""
+    Resource with a fixed ``capacity`` of usage slots
+
+    A process may :py:meth:`request` a single usage slot, which is granted
+    as soon as it becomes available. When all slots are taken, each further
+    :py:meth:`request` is queued until a previous request is :py:meth:`release`\ d.
+
+    .. warning::
+
+        Both :py:meth:`~.put` and :py:meth:`~.get` cannot be used on this resource type.
+    """
+    def __init__(self, env: Environment, capacity: int):
+        if capacity <= 0:
+            raise ValueError("capacity must be greater than 0")
+        super().__init__(env, capacity)
+        #: All :py:class:`~.Request`\ s currently granted for the resource.
+        self.users = []  # type: List[Request]
+
+    @property
+    def queue(self) -> List[Request]:
+        r"""Pending :py:class:`~.Request`\ s currently waiting for the resource."""
+        return self.put_queue
+
+    @property
+    def count(self) -> int:
+        r"""Number of :py:class:`~.Request`\ s currently granted for the resource."""
+        return len(self.users)
+
+    # These do *not* raise a well-defined error in SimPy,
+    # but instead fail with a follow-up AttributeError.
+    # We raise said error *now* to have a better help message.
+    def put(self):
+        raise AttributeError(
+            f"cannot put/get {self.__class__.__name__}, use request/release instead"
+        )
+
+    def get(self):
+        raise AttributeError(
+            f"cannot put/get {self.__class__.__name__}, use request/release instead"
+        )
+
+    def request(self) -> Request:
+        """Request usage of a ``resource``"""
+        return Request(self)
+
+    def release(self, request: Request) -> Release:
+        """Release a previous usage ``request`` of a ``resource``"""
+        return Release(self, request)
+
+    def _do_put(self, event: Request) -> bool:
+        if len(self.users) < self._capacity:
+            self.users.append(event)
+            event.usage_since = self._env.now
+            event.succeed()
+            return True
+        return False
+
+    def _do_get(self, event: Release) -> bool:
+        # releasing a Request should be idempotent
+        try:
+            self.users.remove(event.request)
+        except ValueError:
+            pass
+        event.succeed()
+        # releasing always succeeds
+        return True
```

Comment:
There is basically no condition where releasing a request would not succeed: If it has been granted, it can be released (the maximum and minimum count are fixed). If it hasn't been granted, we just drop.